### DISCUSSION TITLE
Dealing with multiple requests in one question

### DISCUSSION THREAD

**Discussion by simonhackler** [bromberry-games/Inkvestigations#136] [Game design]
Players sometimes add multiple request to a prompt.
E.g. Interrogate the suspects and search the crime scene. 
This will lead to bad responses for both questions since gpt will try to answer all the questions in one.
I think there are multiple ways to deal with this, either in the brain prompt or by parsing the prompt before passing it to the brain.

### 1. Block in the prompt.
Create a few shot example that just "blocks" multiple requests immedieately.
"I am sorry sherlock but I can only do one thing, what would you like me to do?"
#### Advantages:
- Easy fix?
#### Disadvantages: 
- Really annoying for the player
- takes you out of the game immedieately. Why can you do only one thing!!?!?

### 2. Only do one thing
Create a few shot example that takes a multiple request and then outputs only the answer to one.
Then add: Sorry sherlock can only do one thing per day.
#### Advantages:
- is natural enough
- easy fix?
#### Disadvantages:
- "dirties" the few shot prompts even more
- might steer the model off track

### 3. Parse prompt before inputting
Use gpt to parse the prompt before going into the brain. So it will just parse out all the requests and then input the first request to the brain model. Then just add to the letter, sorry holmes can only do one thing per day.
#### Advantages:
- keeps the brain prompt clean
#### Disadvantages:
- slower
- more complex
- more tokens
- more work to implement

@zivky you have another idea?
I think we should try the 2. but I really would prefer the 3. This is something we have to do eventually anyways for saving and loading from the db later when we use gpt-4.
Also we can then have some really nice few shots for the brain because we know exactly how the input format would look.
That would also be a huge advantage.

**Comment by zivky** [comment]
I really don't have anything apart from what you've written. That is: prompt it to handle it. 
As you said, blocking is really annoying and if I understand correctly it would waste a message, which would make it infinitely more annoying. Also, I don't trust gpt3.5 to handle it well with one fewshot. For example, if I say "take the book AND find this passage" I am convinced that it sometimes would give you a negative response even though it makes sense. So nothing gets done in exchange for one message.  
While it definitely also sounds annoying to only get half the answer, I think it isn't as frustrating. So in the same example, Wellington would at least take the book. I'm not sure it can be done with one fewshot though. Will try it of course.

#3 sounds very sophisticated and sounds like it would take the whole game to the next level. But yea, tokens, time, and tinkering. 

So for now, I'll try to work it into the prompt. Perhaps it doesn't need to be one of the fewshots.

**Reply by simonhackler** [reply]
Oh yeah that is true. Can be really annoying in cases where you will have to describe actions that belong together.

It is definitely 3. then in the long term but 2. in the short term? We just try 2. for a bit and if it does not work leave it like that and eventually add 3.

**Reply by zivky** [reply]
A quick try of inserting a special rule and two few-shots concerning two actions at once, doesn't really change how GPT responds, but the answers are pretty solid actually, even when mixing one order that it knows and one it doesn't. I'll try it more thoroughly later. 
Edit: I'm also worried about the size of the individual generations if they have to account for two orders at once. The longer it has to reason the higher the chance it'll break.
